File: app/src/main/java/com/games/memorygame/ui/BasePresenter.java
package com.games.memorygame.ui;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;


public abstract class BasePresenter<T extends MvpView> implements Presenter<T> {

    private T view;
    private CompositeDisposable disposables = new CompositeDisposable();

    @Override
    public void attachView(T view) {
        this.view = view;
    }

    @Override
    public void detachView() {
        this.view = null;
        disposables.clear();
    }

    public T getView() {
        return this.view;
    }

    protected void addDisposable(Disposable disposable) {
        disposables.add(disposable);
    }

    protected void assertViewAttached() {
        if (!this.isViewAttached()) {
            throw new ViewNotAttachedException();
        }
    }

    private boolean isViewAttached() {
        return view != null;
    }

    public static class ViewNotAttachedException extends RuntimeException {
        public ViewNotAttachedException() {
            super("Call Presenter.attachView(MvpView)");
        }
    }
}

File: app/src/main/java/com/games/memorygame/ui/BasePresenter.kt
package com.games.memorygame.ui

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable


abstract class BasePresenter<T : MvpView> : Presenter<T> {

    var view: T? = null
        private set
    private val disposables = CompositeDisposable()

    private val isViewAttached: Boolean
        get() = view != null

    override fun attachView(view: T) {
        this.view = view
    }

    override fun detachView() {
        this.view = null
        disposables.clear()
    }

    protected fun addDisposable(disposable: Disposable) {
        disposables.add(disposable)
    }

    protected fun assertViewAttached() {
        if (!this.isViewAttached) throw ViewNotAttachedException()
    }

    class ViewNotAttachedException : RuntimeException("Call Presenter.attachView(MvpView)")
}
